edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/MetricFilter.cs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.Diagnostics
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// This class acts as a pass through for a group of metrics.
+    /// It will exclude metrics based on a tag whitelist and blacklist.
+    /// After filtering, it will add or remove tags.
+    /// </summary>
+    public class MetricFilter
+    {
+        Option<List<KeyValuePair<string, string>>> tagsWhitelist = Option.None<List<KeyValuePair<string, string>>>();
+        Option<List<KeyValuePair<string, string>>> tagsToAdd = Option.None<List<KeyValuePair<string, string>>>();
+        Option<List<string>> tagsToRemove = Option.None<List<string>>();
+
+        public IEnumerable<Metric> FilterMetrics(IEnumerable<Metric> metrics)
+        {
+            foreach (Metric metric in metrics)
+            {
+                // Skip metric if the whitelist does not contain it.
+                if (this.tagsWhitelist.Exists(wl => !wl.Any(metric.Tags.Contains)))
+                {
+                    continue;
+                }
+
+                // Add or remove tags if needed.
+                if (this.tagsToAdd.HasValue || this.tagsToRemove.HasValue)
+                {
+                    Dictionary<string, string> newTags = metric.Tags.ToDictionary(t => t.Key, t => t.Value);
+
+                    this.tagsToAdd.ForEach(tta => tta.ForEach(toAdd => newTags.Add(toAdd.Key, toAdd.Value)));
+                    this.tagsToRemove.ForEach(ttr => ttr.ForEach(toRemove => newTags.Remove(toRemove)));
+
+                    yield return new Metric(metric.TimeGeneratedUtc, metric.Name, metric.Value, newTags);
+                }
+                else
+                {
+                    yield return metric;
+                }
+            }
+        }
+
+        public MetricFilter AddToWhitelist(params KeyValuePair<string, string>[] pairs)
```

Comment:
I don't see any place where you make use of the params keyword... why not just take a list of KVPairs?

Reply:
This applies to all functions here.

Reply:
Currently where this is used is 
```
this.metricFilter = new MetricFilter()
    .AddToWhitelist(new KeyValuePair<string, string>(MetricsConstants.MsTelemetry, true.ToString()))
    .AddTagsToRemove(MetricsConstants.MsTelemetry, MetricsConstants.IotHubLabel, MetricsConstants.DeviceIdLabel);
```

So both are using the params functionality. Currently, whitelist is only 1 value, but it makes sense to use params in case that grows in the future. It also keeps the signatures consistent.